jsonpb/jsonpb.go

```diff
@@ -644,7 +649,105 @@ func (u *Unmarshaler) UnmarshalNext(dec *json.Decoder, pb proto.Message) error {
 // permutations of the related Marshaler.
 func (u *Unmarshaler) Unmarshal(r io.Reader, pb proto.Message) error {
 	dec := json.NewDecoder(r)
-	return u.UnmarshalNext(dec, pb)
+	if err := u.UnmarshalNext(dec, pb); err != nil {
+		return err
+	}
+
+	return checkRequiredFields(pb)
+}
+
+// checkRequiredFields returns an error if any required field in the given proto message is not set.
+// This function is called from Unmarshal and assumes certain actions are done ahead.
+func checkRequiredFields(pb proto.Message) error {
+	// Most well-known type messages do not contain any required field.  The "Any" type may have
+	// required fields, however the Unmarshaler should have invoked proto.Marshal on the value field
+	// during unmarshaling and that would have returned an error if a required field is not set.
+	if _, ok := pb.(wkt); ok {
+		return nil
+	}
+
+	v := reflect.ValueOf(pb).Elem()
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		sfield := v.Type().Field(i)
+		if strings.HasPrefix(sfield.Name, "XXX_") {
+			continue
+		}
+
+		// Oneof fields need special handling.
+		if sfield.Tag.Get("protobuf_oneof") != "" {
+			// field is an interface containing &T{real_value}.
+			v := field.Elem().Elem() // interface -> *T -> T
+			field = v.Field(0)
+			sfield = v.Type().Field(0)
+		}
+
+		var prop proto.Properties
+		prop.Init(sfield.Type, sfield.Name, sfield.Tag.Get("protobuf"), &sfield)
+
+		// IsNil will panic on most value kinds.
+		switch field.Kind() {
+		case reflect.Map:
```

Comment:
I think we discussed this before, and you told me how a proto3 could contain a proto2 message, but maps are exclusive to proto3, so do they really need to be checked?

Consider adding some code comments explaining the rational of checking maps despite their appearance being expected to be mutually exclusive with required fields.

Reply:
No, maps are allowed in proto2.  I have a test case for this.